filemanager/gcsmanager.go

```diff
@@ -85,18 +101,31 @@ func (m *gcsManager) Upload(ctx context.Context, file *os.File, prefixes ...stri
 	ctx, cancel := context.WithTimeout(ctx, m.getTimeout())
 	defer cancel()
 
-	obj := client.Bucket(m.config.Bucket).Object(fileName)
-	w := obj.NewWriter(ctx)
-	if _, err := io.Copy(w, file); err != nil {
-		err = fmt.Errorf("copying file to GCS: %v", err)
-		if closeErr := w.Close(); closeErr != nil {
-			return UploadedFile{}, fmt.Errorf("closing writer: %q, while: %w", closeErr, err)
-		}
+	object := client.Bucket(m.config.Bucket).Object(fileName)
+	if m.config.doNotOverWrite {
+		object = object.If(storage.Conditions{DoesNotExist: true})
+	}
+	w := object.NewWriter(ctx)
 
-		return UploadedFile{}, err
+	if _, err := io.Copy(w, file); err != nil {
+		return UploadedFile{}, fmt.Errorf("copying file to writer: %v", err)
 	}
 
 	if err := w.Close(); err != nil {
+		switch e := err.(type) {
+		case *googleapi.Error:
+			if e.Code == http.StatusPreconditionFailed {
+				m.logger.Warnn(
+					"object already exists",
+					logger.NewStringField("object", fileName),
+				)
+				return UploadedFile{
+					Location:   "not uploaded, already exists",
+					ObjectName: fileName,
+				}, nil
+			}
+		default:
+		}
```

Comment:
It might fail if the error is wrapped.
```suggestion
		var ge *googleapi.Error
		switch {
		case errors.As(err, &ge):
			if ge.Code == http.StatusPreconditionFailed {
				m.logger.Warnn(
					"object already exists",
					logger.NewStringField("object", fileName),
				)
				return UploadedFile{
					Location:   "not uploaded, already exists",
					ObjectName: fileName,
				}, nil
			}
		}
```

Reply:
Didn't get you. It's an error returned by the gcs(`storage`) package, we don't wrap it either.

Reply:
It's more safer to use `errors.As` since `Type switch on errors fails on wrapped errors` and errors returned by the gcs(storage) package can we wrapped.

Reply:
Understood.
I've done away with the switch case in that case(considering we're only interested in one status code)
```
        if err := w.Close(); err != nil {
		var ge *googleapi.Error
		if errors.As(err, &ge) && ge.Code == http.StatusPreconditionFailed {
			return UploadedFile{}, ErrPreConditionFailed
		}
		return UploadedFile{}, fmt.Errorf("closing writer: %w", err)
	}
```